node/src/store/utils.rs

```diff
@@ -13,52 +14,42 @@ use crate::store::{SamplingMetadata, StoreError};
 pub(crate) const VALIDATIONS_PER_YIELD: usize = 4;
 
 /// based on the stored headers and current network head height, calculate range of headers that
-/// should be fetched from the network, starting from the front up to a `limit` of headers
+/// should be fetched from the network, anchored on already existing header range in store
 pub(crate) fn calculate_range_to_fetch(
-    head_height: u64,
+    subjective_head_height: u64,
     store_headers: &[RangeInclusive<u64>],
-    syncing_window_edge: Option<u64>,
+    //syncing_window_edge: Option<u64>,
     limit: u64,
 ) -> BlockRange {
-    let mut missing_range = get_most_recent_missing_range(head_height, store_headers);
-
-    // truncate to syncing window, if height is known
-    if let Some(window_edge) = syncing_window_edge {
-        if missing_range.start() < &window_edge {
-            missing_range = window_edge + 1..=*missing_range.end();
-        }
-    }
-
-    // truncate number of headers to limit
-    if missing_range.len() > limit {
-        let end = missing_range.end();
-        let start = end.saturating_sub(limit) + 1;
-        missing_range = start..=*end;
+    match get_most_recent_missing_range(subjective_head_height, store_headers) {
+        Either::Left(range) => range.truncate_left(limit),
+        Either::Right(range) => range.truncate_right(limit),
     }
-
-    missing_range
 }
 
+/// Return next range that should be downloaded, `Either::Left` of existing range
+/// (syncing backwards, should be trimmed from the left to extend existing range correctly)
+/// or `Either::Right` of stored ranges (syncing forward, trimming from the right)
 fn get_most_recent_missing_range(
     head_height: u64,
     store_headers: &[RangeInclusive<u64>],
-) -> BlockRange {
+) -> Either<BlockRange, BlockRange> {
     let mut store_headers_iter = store_headers.iter().rev();
 
     let Some(store_head_range) = store_headers_iter.next() else {
         // empty store, we're missing everything
-        return 1..=head_height;
+        return Either::Right(1..=head_height);
     };
 
     if store_head_range.end() < &head_height {
-        // if we haven't caught up with network head, start from there
-        return store_head_range.end() + 1..=head_height;
+        // if we haven't caught up with the network head, start from there
+        return Either::Right(store_head_range.end() + 1..=head_height);
     }
 
     // there exists a range contiguous with network head. inspect previous range end
     let penultimate_range_end = store_headers_iter.next().map(|r| *r.end()).unwrap_or(0);
 
-    penultimate_range_end + 1..=store_head_range.start().saturating_sub(1)
+    Either::Left(penultimate_range_end + 1..=store_head_range.start().saturating_sub(1))
 }
```

Comment:
Instead of using `Either`, it makes sense to merge this into one function. For example `Either::Right(1..=head_height)` can become `(1..=head_height).truncate_right(limit)`

Something like this:

```rust
pub(crate) fn calculate_range_to_fetch(
    subjective_head_height: u64,
    store_headers: &[RangeInclusive<u64>],
    limit: u64,
) -> BlockRange {
    let mut store_headers_iter = store_headers.iter().rev();

    let Some(store_head_range) = store_headers_iter.next() else {
        // empty store, we're missing everything
        let range = 1..=subjective_head_height;
        return range.truncate_right(limit);
    };

    if *store_head_range.end() < subjective_head_height {
        // if we haven't caught up with the network head, start from there
        let range = store_head_range.end() + 1..=subjective_head_height;
        return range.truncate_right(limit);
    }

    // there exists a range contiguous with network head. inspect previous range end
    let penultimate_range_end = store_headers_iter.next().map(|r| *r.end()).unwrap_or(0);

    let range = penultimate_range_end + 1..=store_head_range.start().saturating_sub(1);
    range.truncate_left(limit)
}
```

Also I prefer this helper function to be in `syncer.rs`

Reply:
I think you're right, now it's simple enough that it doesn't make sense to split this into two functions. Merged and moved to syncer :+1: